[Android] 페이지 계산 시점에 폰트가 로딩되지 않아서 오차가 발생할 수 있는 문제 수정

kgwangrae opened the pull request:
우선 `document.fonts`를 사용할 수 있는 경우에 대해서만 대응하였습니다.

그 외의 경우(예를 들어 Android 4.4 버전 이하의 기기)에 대해서는 아래의 두 가지 방법을 확인해보았지만, 문제가 있어 대응을 보류하였습니다.
- `document.readyState === 'complete'` 인지 확인하는 방법
  - Chrome 62에서 `document.fonts.status`의 결과와 대조해보니 fontface를 여전히 불러오고 있는데도 위 값이 `true`가 반환되는 경우가 있는 것을 확인하였습니다. 즉, 신뢰할 수 없습니다.
- https://github.com/bramstein/fontfaceobserver 
  - Observing할 font family를 개별적으로 지정해주어야하는 단점이 있습니다.
  - `document.body`에 text element를 추가한 후, 폰트가 로딩됨에 따라 해당 element의 width가 변화하는지 추적하는 방식이라 DOM에 영향을 줍니다.

Commit:
[안드(5.0이상)] 페이지 계산 시 폰트가 로딩되지 않아서 오차가 발생할 수 있는 문제 수정

- 안드 4.4 이하는 추후 대응 필요

DavinAhn commented on src/android/Reader.es6:
```diff
@@ -146,9 +146,15 @@ export default class Reader extends _Reader {
    */
   calcPageCount() {
+    if (document.fonts && document.fonts.status) {
+      if (document.fonts.status !== 'loaded') {
```
`status`의 종류에는 무엇이 있나요?
참고한 문서를 코멘트하면 좋겠습니다.

kgwangrae replied:
원래 https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet 문서를 보고 `loading` 아니면 `loaded` 뿐이라고 생각했는데 아니었네요 😰 

https://drafts.csswg.org/css-font-loading/ 에 의하면 아래와 같습니다.

```
enum FontFaceLoadStatus { "unloaded", "loading", "loaded", "error" };
```

이에 따르면 조건을 `document.fonts.status === 'unloaded' || document.fonts.status === 'loading'` 으로 변경하는 것이 더 좋겠군요;

Commit:
b501461c9 의 방법 대신 조금 더 신뢰할 수 있는 방법으로 바꿈

- FontFace가 여러 개가 있고 각자 loading status가 다를 수 있다.
- 이때 document.fonts.status가 항상 loading이나 loaded 만을 반환한다. (심지어 모든 FontFace 로딩에 실패해도 loaded가 반환될 수 있다.)
- 신뢰할 수 없는 document.fonts.status 대신 개별 FontFace의 상태를 체크하도록 함

DavinAhn commented on src/android/Reader.es6:
```diff
@@ -146,9 +146,23 @@ export default class Reader extends _Reader {
+      // 로딩 실패한 Font : error
+      // document.fonts.status, ready는 신뢰할 수 없으므로 아래와 같은 방법으로 체크
+      const fontFaceLoadingStatusList =
+        Array.from(document.fonts.values()).map(fontFace => fontFace.status);
```
`Array.from`은 `document.fonts`보다 이후에 추가된 메소드입니다.
`[].slice.call`로 대체하거나 polyfill을 추가(비권장)해주세요.

kgwangrae replied:
수정했습니다.

DavinAhn commented on src/android/Reader.es6:
```diff
@@ -146,9 +146,23 @@ export default class Reader extends _Reader {
+      // document.fonts.status, ready는 신뢰할 수 없으므로 아래와 같은 방법으로 체크
+      const fontFaceLoadingStatusList =
+        Array.from(document.fonts.values()).map(fontFace => fontFace.status);
+      if (fontFaceLoadingStatusList.indexOf('loading') >= 0) {
```
외부 링크에 의한 폰트 로딩이 이뤄질 경우 네트워크 지연 등의 이유로 인해 로딩이 길어지면(ex: timeout) 어떻게 되나요?

kgwangrae replied:
Mac용 Chrome 67에서 웹 폰트가 다수 포함되어있는 https://fonts.google.com/ 페이지에 접속하여 Slow 3G 환경으로 Network Throttling 설정해두고 테스트를 해보았습니다. 위 코드가 별 문제 없이 동작할 것으로 보입니다 😃 

```
document.fonts.onloading = (e) => { console.log(new Date()); console.log(e); console.log(Array.from(document.fonts.values()).map((elem) => elem.status).join()); }

document.fonts.onloadingdone = (e) => { console.log(new Date()); console.log(e); console.log(Array.from(document.fonts.values()).map((elem) => elem.status).join()); }

// Scroll down

(2) java/com/google/fonts/directory/ui/angular_js.js:132 Loading fonts (20)

VM1337:1 Mon Jun 25 2018 12:13:52 GMT+0900 (한국 표준시)
VM1337:1 FontFaceSetLoadEvent {isTrusted: true, fontfaces: Array(0), type: "loading", target: FontFaceSet, currentTarget: FontFaceSet, …}
VM1337:1 loaded,unloaded,loaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,loaded,unloaded,unloaded,unloaded,unloaded,unloaded,loaded,loaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,unloaded,unloaded,unloaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading,loading
VM1340:1 Mon Jun 25 2018 12:13:58 GMT+0900 (한국 표준시)
VM1340:1 FontFaceSetLoadEvent {isTrusted: true, fontfaces: Array(40), type: "loadingdone", target: FontFaceSet, currentTarget: FontFaceSet, …}
VM1340:1 loaded,unloaded,loaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,loaded,unloaded,unloaded,unloaded,unloaded,unloaded,loaded,loaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,unloaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,unloaded,unloaded,unloaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded,loaded
```

DavinAhn replied:
좋네요!

Commit:
Array.from 이용하지 않도록 함

- 호환성 문제 발생할 수 있음
